### ISSUE TITLE
Datadog consumer - F5 telemetry integration

### ISSUE THREAD

**Issue by anesmashnyi-godaddy** [F5Networks/f5-telemetry-streaming#94]
### Is your feature request related to a problem? Please describe.
F5 Telemetry Service currently supports AWS S3, AWS CloudWatch. Would like to have Datadog -  F5 running in AWS added to the push consumers list

### Describe the solution you'd like
Datadog push consumer for F5 in AWS.

### Describe alternatives you've considered
None

### Additional context
None

**Comment by dstokesf5** [comment]
This request is currently in our internal product backlog as AUTOTOOL-622.

**Comment by dstokesf5** [comment]
We are still looking into a dedicated Datadog consumer. In the meantime, it is possible to send data to Datadog using existing consumers.

You can send log data to the web API using the Generic_HTTP consumer: https://docs.datadoghq.com/api/latest/logs/#send-logs
You can send metrics to a running Datadog agent by using the StatsD consumer: https://docs.datadoghq.com/developers/dogstatsd/?tab=hostagent

Below is an example declaration to send logs and metrics to Datadog with existing consumers:
```json
{
    "class": "Telemetry",
    "datadogMetrics": {
        "class": "Telemetry_Namespace",
        "system": {
            "class": "Telemetry_System",
            "systemPoller": {
                "interval": 60
            }
        },
        "dogStatsDConsumer": {
            "class": "Telemetry_Consumer",
            "type": "Statsd",
            "host": "{{dogstatsdAgentAddress}}",
            "protocol": "udp",
            "port": 8125
        }
    },
    "datadogLogs": {
        "class": "Telemetry_Namespace",
        "eventListener": {
            "class": "Telemetry_Listener",
            "port": 6514
        },
        "datadogLogConsumer": {
            "class": "Telemetry_Consumer",
            "type": "Generic_HTTP",
            "host": "http-intake.logs.datadoghq.com",
            "path": "/v1/input",
            "headers": [
                {
                    "name": "content-type",
                    "value": "application/json"
                },
                {
                    "name": "DD-API-KEY",
                    "value": "{{datadogApiKey}}"
                }
            ]
        }
    }
}
```

**Comment by amolari** [comment]
@dstokesf5 
As far as I've seen:
Issues with dogStatsDconsumer:
- array metrics will be stripped and not transformed (as in AWS CloudWatch consumer), so the number of metrics is quite reduced
- issue with dimensions: the Statsd consumer seems to use "etsy version" of Statsd, which is quite basic and does not support what newer statsd "versions" (Telegraf, dogStatsd) have: tag support.

Issues with datadogLogConsumer:
- the datadog API expect the json to be formatted differently https://github.com/F5Networks/f5-telemetry-streaming/issues/114 :
   * "message" instead of "data"
   * "service" instead of "telemetryEventCategory"
   * "hostname" instead of "host"
- no possibility to add/configure new keys used by datadog API such as "ddsource" and "ddtags"
- timestamp is not iso formatted https://github.com/F5Networks/f5-telemetry-streaming/issues/117

**Comment by dstokesf5** [comment]
I very much appreciate the additional feedback. This will help us not only improve those existing consumers, but will help us in creating a better dedicated Datadog consumer. I am working on adding additional items to our product backlog based on this feedback.

**Comment by amolari** [comment]
@dstokesf5 
One very important point I beleive, is the support of custom tags (custom key/custom value):
- be able to provide to Datadog all the non purely metric (non numerical values) data. There are quite a lot of them that are required, as the RestAPI in many occasions only have the stats/info in string format only (vs snmp which (usually?) propose both format: OID returning numerical value => metric and for the sam stat an OID returning a string).
- talking to our Datadog specialist, it seems that Datadog has no known limit in tag key (or value) length, and number of tags.

**Comment by dstokesf5** [comment]
@amolari Are you requesting that all non-metric data get automatically added to the StatsD output as tags? Or are you requesting something similar to the custom dimensions feature request where dynamic data can be specified to be included as tags?

We are currently looking at using [JMESPath](https://jmespath.org/) for user defined payload modifications on the Generic HTTP consumer to help with the limitations of datadogLogConsumer from the example I provided. I am thinking JMESPath could also be used to specify dynamic tags and dimensions. Perhaps with something like this:
```json
{
    "class": "Telemetry_Consumer",
    "type": "Statsd",
    "host": "192.0.2.0",
    "protocol": "udp",
    "tags": {
        "hostname": "system.hostname",
        "version": "system.version",
        "versionBuild": "system.versionBuild",
        "staticString": "'email@example.com'"
    },
    "dimensions": [
        "'f5-telemetry'",
        "system.location"
    ],
    "port": 8125
}
```
I would appreciate any feedback you have on using JMESPath to define custom tags and dimensions.

**Comment by dstokesf5** [comment]
@amolari For reporting the status of BIG-IP items, such as virtuals, pools, and pool members, we could look into supporting DataDog's service checks: https://docs.datadoghq.com/developers/service_checks/

However, this is only helpful for one use case of the string data found in metrics. Are there other use cases you have for string metrics that we did not discuss in our meeting or that I am already forgetting?

**Comment by dstokesf5** [comment]
I have been giving some more thought to the idea of automatically including string metrics as tags. If we automatically include all string metrics from a system poll on each individual numeric metric, I am worried we would make the data difficult to work with.

Instead, I am wondering if it would be sufficient to include all string metrics as tags for any metrics adjacent to the current numerical metric. I think this is best described with an example. Below is some example system poller output:
```json
...
    "virtualServers": {
        "/Common/app.app/app_vs": {
            "clientside.bitsIn": 0,
            "clientside.bitsOut": 0,
            "clientside.curConns": 0,
            "clientside.maxConns": 0,
            "clientside.pktsIn": 0,
            "clientside.pktsOut": 0,
            "clientside.totConns": 0,
            "destination": "10.5.6.7:80",
            "availabilityState": "offline",
            "enabledState": "enabled",
            "status.statusReason": "The virtual server is available",
            "mask": "255.255.255.255",
            "name": "/Common/foofoo.app/foofoo_vs",
            "appService": "/Common/foofoo.app/foofoo",
            "ipProtocol": "tcp",
            "tenant": "Common",
            "pool": "/Common/foofoo.app/foofoo_pool",
            "application": "foofoo.app",
            "profiles": {
                "/Common/tcp": {
                    "name": "/Common/tcp",
                    "tenant": "Common"
                },
                "/Common/app/http": {
                    "name": "/Common/app/http",
                    "tenant": "Common",
                    "application": "app"
                }
            }

        },
...
```
I am thinking that for each numeric metric (e.g. clientside.bitsIn, clientsid.totConns, etc.) we could include all of the following data points as tags:
* destination
* availabilityState
* enabledState
* status.statusReason
* mask
* name
* appService
* ipProtocol
* tenant
* pool
* application

**Comment by amolari** [comment]
> @amolari For reporting the status of BIG-IP items, such as virtuals, pools, and pool members, we could look into supporting DataDog's service checks: https://docs.datadoghq.com/developers/service_checks/
> 
> However, this is only helpful for one use case of the string data found in metrics. Are there other use cases you have for string metrics that we did not discuss in our meeting or that I am already forgetting?

The following are the most important ones (1st prio), from my perspective:
- Category "system"
      * hostname (must be always sent in the mandatory datadog tag "hostname" anyway)
      *  version
      *  versionBuild
      *  baseMac
      *  syncStatus
      *  failoverStatus
      * asmState
- Category "networkInterfaces"
      *  status
- Category "virtualServers"
      * availabilityState
      * enabledState
      * name

**Comment by jmcalalang** [comment]
Bump for tracking and update on if this is coming. VIU

**Comment by dstokesf5** [comment]
@jmcalalang We are looking into getting a dedicated DataDog consumer out in the 1.22.0 release. If you are fine with using a DataDog agent, you should be able to use the StatsD and Generic_HTTP consumers in the meantime.

**Comment by dstokesf5** [comment]
This issue has been resolved with the 1.22.0 release of Telemetry Streaming.
